lib/registry/registry-client.ts

```diff
@@ -0,0 +1,621 @@
+/**
+ * @license
+ * Copyright 2017 Balena Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Attribution:  Guided by code at https://github.com/joyent/node-docker-registry-client
```

Comment:
License-wise, are we OK to publish this code in `balena-multibuild` under Apache v2.0? It looks like the joyent code is licensed under Mozilla Public License Version 2.0, but it was itself and adaptation of code from elsewhere:
```
Adapted from
 * <github.com/mcavage/node-restify/blob/master/lib/clients/string_client.js>
 * now at <https://github.com/restify/clients/blob/master/lib/StringClient.js>
```
And the latter seems to be covered by the MIT license. Depending on how similar (is it a copy?) the code in `balena-multibuild` is to code in those other repos, we would have to comply with their licenses.

Reply:
I thought a long time about this.  I'm not sure about the answer.  Originally I started with copying code from joyent but honestly it was so crufty that I finally blew it away and basically re-wrote the code using it as a guide.  To me, the two are fairly different at this point.

Reply:
> To me, the two are fairly different at this point.

I understand that original copyright licenses apply if the original code and your rewriting are "substantially similar". Would someone reading your code and the original find them them to be "substantially similar"? (Merely porting from JS to TS or vice versa would count as substantially similar.) I understand the "substantially similar" rule is the same as for books. If you read a story and the retell it / rewrite it _"using your own words,"_ such that sentences and paragraphs will be completely different from the original, I understand this does not count as "substantially similar" and therefore copyright licensing terms do not apply. By contrast, if you copy and paste a story and systematically change "a third of the words" while maintaining the same grammatical structure and sequence of sentences, this will be substantially similar and copyright licensing applies.

If your code cannot be said to be "substantially similar" to the original, then the original licensing does not apply. If this is the case, I would advise deleting the _"Attribution:  Guided by code at"_ comment, especially as added to licensing header.

Reply:
The code was not just ported over from JS to TS.  In my estimation they are not substantially similar, but I'm not an expert.  I'll go with that and delete the comment.